File: frontend/src/lib/economics.ts
import type { Assessment, Inputs } from '@/types';
import { ceilUnits, div, money, mul, parse } from '@/lib/money';

/** Two formula versions exist on purpose (action plan T04).
 *
 *  `unit-economics/1.0.0` is the float implementation the pilot shipped. It disagreed
 *  with the server on some accepted inputs, so nothing new is calculated with it — but
 *  assessments already saved under it must still replay to the values they were saved
 *  with, so `calculateV1_0_0` stays.
 *
 *  `unit-economics/1.1.0` is the current version: identical formulas, evaluated with the
 *  shared scaled-integer arithmetic in `money.ts` so the browser and the API agree exactly. */
export const FORMULA_VERSION = 'unit-economics/1.1.0';
export const THRESHOLD_VERSION = 'decision-gates/1.0.0';

export interface EvidenceGate { confidence: number; coverage: boolean; compliance_resolved: boolean; overall: number | null; observation_ids: string[] }
export const NO_EVIDENCE: EvidenceGate = { confidence: 0, coverage: false, compliance_resolved: false, overall: null, observation_ids: [] };

type ScenarioRow = Assessment['scenarios'][number];

/** Legacy (1.0.0) rounding. Kept only so stored assessments replay unchanged. */
const legacyMoney = (value: number) => Math.sign(value) * Math.round((Math.abs(value) + Number.EPSILON) * 100) / 100;

function scenariosV1_0_0(i: Inputs): ScenarioRow[] {
  const factors: [string, number, number][] = [['Downside', 1-i.stress_pct/100, 1+i.stress_pct/100], ['Base', 1, 1], ['Upside', 1+i.stress_pct/100, 1-i.stress_pct/100]];
  return factors.map(([name, priceFactor, costFactor]) => {
    const supplier = i.unit_cost_usd*i.fx_ngn*costFactor, freight = i.freight_ngn/i.quantity*costFactor;
    const duty = (supplier+freight)*i.duty_pct/100, importTax = (supplier+freight+duty)*i.import_tax_pct/100;
    const landed = supplier+freight+duty+importTax, price = i.selling_price_ngn*priceFactor;
    const fees = price*i.channel_fee_pct/100, returns = price*i.returns_pct/100, marketing = i.marketing_ngn/i.quantity, overhead = i.fixed_cost_ngn/i.quantity;
    const contribution = price-landed-fees-returns-marketing-overhead, beforeFixed = price-landed-fees-returns;
    return { name, supplier: legacyMoney(supplier), freight: legacyMoney(freight), duty: legacyMoney(duty), import_tax: legacyMoney(importTax), landed_cost: legacyMoney(landed), price: legacyMoney(price), fees: legacyMoney(fees), returns: legacyMoney(returns), marketing: legacyMoney(marketing), overhead: legacyMoney(overhead), contribution: legacyMoney(contribution), margin_pct: legacyMoney(contribution/price*100), cash_required: legacyMoney(landed*i.quantity+i.marketing_ngn+i.fixed_cost_ngn), break_even_cac: legacyMoney(Math.max(0,beforeFixed-overhead)), break_even_units: beforeFixed>0 ? Math.ceil((i.marketing_ngn+i.fixed_cost_ngn)/beforeFixed) : null };
  });
}

/** The same formulas on scaled integers, so the server can match exactly. */
function scenariosV1_1_0(i: Inputs): ScenarioRow[] {
  const one = parse(1), hundred = parse(100);
  const quantity = parse(i.quantity);
  const unitCost = parse(i.unit_cost_usd), fx = parse(i.fx_ngn);
  const freightTotal = parse(i.freight_ngn);
  const dutyRate = div(parse(i.duty_pct), hundred), taxRate = div(parse(i.import_tax_pct), hundred);
  const sellingPrice = parse(i.selling_price_ngn);
  const feeRate = div(parse(i.channel_fee_pct), hundred), returnsRate = div(parse(i.returns_pct), hundred);
  const marketingTotal = parse(i.marketing_ngn), fixedTotal = parse(i.fixed_cost_ngn);
  const stress = div(parse(i.stress_pct), hundred);
  const factors: [string, bigint, bigint][] = [['Downside', one-stress, one+stress], ['Base', one, one], ['Upside', one+stress, one-stress]];
  return factors.map(([name, priceFactor, costFactor]) => {
    const supplier = mul(mul(unitCost, fx), costFactor);
    const freight = mul(div(freightTotal, quantity), costFactor);
    const duty = mul(supplier+freight, dutyRate);
    const importTax = mul(supplier+freight+duty, taxRate);
    const landed = supplier+freight+duty+importTax;
    const price = mul(sellingPrice, priceFactor);
    const fees = mul(price, feeRate), returns = mul(price, returnsRate);
    const marketing = div(marketingTotal, quantity), overhead = div(fixedTotal, quantity);
    const contribution = price-landed-fees-returns-marketing-overhead;
    const beforeFixed = price-landed-fees-returns;
    return { name, supplier: money(supplier), freight: money(freight), duty: money(duty), import_tax: money(importTax), landed_cost: money(landed), price: money(price), fees: money(fees), returns: money(returns), marketing: money(marketing), overhead: money(overhead), contribution: money(contribution), margin_pct: money(mul(div(contribution, price), hundred)), cash_required: money(mul(landed, quantity)+marketingTotal+fixedTotal), break_even_cac: money(beforeFixed-overhead > 0n ? beforeFixed-overhead : 0n), break_even_units: beforeFixed > 0n ? ceilUnits(div(marketingTotal+fixedTotal, beforeFixed)) : null };
  });
}

const SCENARIO_FORMULAS: Record<string, (i: Inputs) => ScenarioRow[]> = {
  'unit-economics/1.0.0': scenariosV1_0_0,
  'unit-economics/1.1.0': scenariosV1_1_0,
};

/** The decision gates. Unchanged by T04 and versioned separately as THRESHOLD_VERSION. */
export function gates(scenarios: ScenarioRow[], i: Inputs, evidence: EvidenceGate): { decision: Assessment['decision']; blockers: string[] } {
  const base = scenarios[1], downside = scenarios[0], blockers: string[] = [];
  if (!evidence.coverage) blockers.push('Collect independent demand signals and destination-market evidence.');
  if (!evidence.compliance_resolved) blockers.push('Obtain a reviewed product classification and current import requirements.');
  if (evidence.confidence<70) blockers.push('Raise evidence confidence to at least 70 before committing inventory.');
  if (base.margin_pct<25) blockers.push('Raise base contribution margin to at least 25%.');
  if (downside.margin_pct<10) blockers.push('Keep downside contribution margin at or above 10%.');
  let decision: Assessment['decision'];
  if(i.compliance==='prohibited') { decision='NO-GO'; blockers.unshift('The product is marked prohibited. Do not proceed.'); }
  else if(evidence.confidence<40 || !evidence.coverage) decision='INSUFFICIENT EVIDENCE';
  else if(base.margin_pct<15 || (evidence.overall!==null && evidence.overall<55)) decision='NO-GO';
  else if(!blockers.length && (evidence.overall||0)>=75) decision='GO';
  else decision='WATCH';
  return { decision, blockers };
}

/** Scenarios and gates for `i`.
 *  `formulaVersion` exists so a stored assessment replays under the version it was saved
 *  with. Callers producing a *new* assessment must leave it at the default. */
export function calculate(i: Inputs, evidence: EvidenceGate = NO_EVIDENCE, formulaVersion: string = FORMULA_VERSION): Assessment {
  const build = SCENARIO_FORMULAS[formulaVersion];
  if (!build) throw new Error(`unknown formula version: ${formulaVersion}`);
  const scenarios = build(i);
  const { decision, blockers } = gates(scenarios, i, evidence);
  return {formula_version:formulaVersion,truth_state:'Calculated',currency:'NGN',market:'NG',decision,confidence:evidence.confidence,observation_ids:evidence.observation_ids,blockers,scenarios,inputs:i,input_truth_state:'User input'};
}

export interface EconomicsSummary { base_margin_pct: number; downside_margin_pct: number; contribution: number; break_even_units: number | null; viable: boolean; failures: string[]; threshold_version: string }
/** Whether the scenario pays for itself, judged without reference to evidence.
 *  Kept out of `calculate()` on purpose: that output is pinned by `formula_version` and a
 *  replayed historical assessment must reproduce byte for byte. This is a separate reading
 *  of the same scenarios so a screen can say "the economics fail under these assumptions"
 *  *and* "demand is unverified" instead of collapsing both into one verdict (T07).
 *  Mirrors economics_summary() in backend/app/economics.py. */
export function economicsSummary(scenarios: ScenarioRow[]): EconomicsSummary {
  const base = scenarios[1], downside = scenarios[0], failures: string[] = [];
  if (base.contribution <= 0) failures.push('The unit does not cover its own costs under these assumptions.');
  if (base.margin_pct < 15) failures.push('Base contribution margin is below the 15% floor.');
  else if (base.margin_pct < 25) failures.push('Base contribution margin is below the 25% target.');
  if (downside.margin_pct < 10) failures.push('Downside contribution margin is below 10%.');
  return { base_margin_pct: base.margin_pct, downside_margin_pct: downside.margin_pct, contribution: base.contribution,
    break_even_units: base.break_even_units, viable: !failures.length, failures, threshold_version: THRESHOLD_VERSION };
}

export type NumericKey = Exclude<keyof Inputs,'compliance'|'channel'|'shipping'>;
const SENSITIVITY_KEYS: NumericKey[] = ['unit_cost_usd','fx_ngn','freight_ngn','duty_pct','import_tax_pct','selling_price_ngn','channel_fee_pct','returns_pct','marketing_ngn','fixed_cost_ngn','quantity'];
export interface Swing { key: NumericKey; low: number; high: number; swing: number }
/** Ranks inputs by how far a ±variation% change moves the base contribution margin. Scenarios, not forecasts. */
export function sensitivity(i: Inputs, variation = 10): Swing[] {
  const shift=(key:NumericKey,factor:number)=>{const value=i[key]*factor;return {...i,[key]:key==='quantity'?Math.max(1,Math.round(value)):value};};
  return SENSITIVITY_KEYS.map(key=>{
    const low=calculate(shift(key,1-variation/100)).scenarios[1].margin_pct, high=calculate(shift(key,1+variation/100)).scenarios[1].margin_pct;
    return {key,low,high,swing:Math.abs(high-low)};
  }).filter(s=>s.swing>0.05).sort((a,b)=>b.swing-a.swing).slice(0,3);
}
export interface Targets { target_margin: number; landed: number; price: number | null; max_landed: number | null; max_unit_cost_usd: number | null }
/** The price and landed cost that would satisfy a target base margin under the current inputs. */
export function targets(i: Inputs, target = 25): Targets {
  const variable=(i.channel_fee_pct+i.returns_pct)/100, perUnitFixed=(i.marketing_ngn+i.fixed_cost_ngn)/i.quantity;
  const importFactor=(1+i.duty_pct/100)*(1+i.import_tax_pct/100);
  const landed=(i.unit_cost_usd*i.fx_ngn+i.freight_ngn/i.quantity)*importFactor;
  const headroom=1-variable-target/100;
  const maxLanded=i.selling_price_ngn*headroom-perUnitFixed;
  const maxUnitCost=(maxLanded/importFactor-i.freight_ngn/i.quantity)/i.fx_ngn;
  return {target_margin:target, landed, price:headroom>0?(landed+perUnitFixed)/headroom:null, max_landed:maxLanded>0?maxLanded:null, max_unit_cost_usd:maxLanded>0&&maxUnitCost>0?maxUnitCost:null};
}

File: backend/app/economics.py
"""Deterministic scenarios. No FX, tariff, sales or confidence is fabricated.

Two formula versions live here on purpose (action plan T04):

``unit-economics/1.0.0``
    The float implementation the pilot shipped. It disagreed with the browser on some
    accepted inputs, so nothing new is calculated with it — but every assessment already
    saved under it must still replay to the values it was saved with, so the code stays.

``unit-economics/1.1.0``
    The current version. Identical formulas, evaluated with the shared scaled-integer
    decimal arithmetic in ``money.py`` so the server and the browser agree exactly.

``calculate()`` uses the current version unless a caller asks for a stored one.
"""
from math import ceil
from decimal import Decimal, ROUND_HALF_UP
from typing import Literal
from pydantic import BaseModel, ConfigDict, Field

from . import money as m

FORMULA_VERSION = 'unit-economics/1.1.0'
THRESHOLD_VERSION = 'decision-gates/1.1.0'


class Inputs(BaseModel):
    model_config = ConfigDict(extra='forbid', allow_inf_nan=False)
    quantity: int = Field(ge=1, le=1000000)
    unit_cost_usd: float = Field(ge=0, le=1000000)
    fx_ngn: float = Field(gt=0, le=1000000)
    freight_ngn: float = Field(ge=0, le=1e12)
    duty_pct: float = Field(ge=0, le=100)
    import_tax_pct: float = Field(ge=0, le=100)
    selling_price_ngn: float = Field(gt=0, le=1e12)
    channel_fee_pct: float = Field(ge=0, le=100)
    returns_pct: float = Field(ge=0, le=100)
    marketing_ngn: float = Field(ge=0, le=1e12)
    fixed_cost_ngn: float = Field(ge=0, le=1e12)
    stress_pct: float = Field(ge=0, le=50)
    compliance: Literal['unresolved', 'prohibited'] = 'unresolved'
    channel: str = Field(default='Direct sales', max_length=100)
    shipping: Literal['Air', 'Sea'] = 'Air'


NO_EVIDENCE = {'confidence': 0, 'coverage': False, 'compliance_resolved': False, 'overall': None,
               'observation_ids': [], 'compliance_status': 'none'}


def money(value):
    """Legacy (1.0.0) rounding: quantise the shortest decimal form of a float."""
    return float(Decimal(str(value)).quantize(Decimal('.01'), rounding=ROUND_HALF_UP))


def _scenarios_v1_0_0(i):
    scenarios = []
    for name, price_factor, cost_factor in [('Downside', 1-i.stress_pct/100, 1+i.stress_pct/100), ('Base', 1, 1), ('Upside', 1+i.stress_pct/100, 1-i.stress_pct/100)]:
        supplier = i.unit_cost_usd * i.fx_ngn * cost_factor
        freight = i.freight_ngn / i.quantity * cost_factor
        duty = (supplier + freight) * i.duty_pct / 100
        tax = (supplier + freight + duty) * i.import_tax_pct / 100
        landed = supplier + freight + duty + tax
        price = i.selling_price_ngn * price_factor
        fees = price * i.channel_fee_pct / 100
        returns = price * i.returns_pct / 100
        marketing = i.marketing_ngn / i.quantity
        overhead = i.fixed_cost_ngn / i.quantity
        contribution = price - landed - fees - returns - marketing - overhead
        before_fixed = price - landed - fees - returns
        scenarios.append({
            'name': name, 'supplier': money(supplier), 'freight': money(freight), 'duty': money(duty),
            'import_tax': money(tax), 'landed_cost': money(landed), 'price': money(price),
            'fees': money(fees), 'returns': money(returns), 'marketing': money(marketing), 'overhead': money(overhead),
            'contribution': money(contribution), 'margin_pct': money(contribution/price*100),
            'cash_required': money(landed*i.quantity + i.marketing_ngn + i.fixed_cost_ngn),
            'break_even_cac': money(max(0, before_fixed-overhead)),
            'break_even_units': ceil((i.marketing_ngn+i.fixed_cost_ngn)/before_fixed) if before_fixed > 0 else None,
        })
    return scenarios


def _scenarios_v1_1_0(i):
    """The same formulas, evaluated on scaled integers so the browser can match exactly."""
    one, hundred = m.parse(1), m.parse(100)
    quantity = m.parse(i.quantity)
    unit_cost = m.parse(i.unit_cost_usd)
    fx = m.parse(i.fx_ngn)
    freight_total = m.parse(i.freight_ngn)
    duty_rate = m.div(m.parse(i.duty_pct), hundred)
    tax_rate = m.div(m.parse(i.import_tax_pct), hundred)
    selling_price = m.parse(i.selling_price_ngn)
    fee_rate = m.div(m.parse(i.channel_fee_pct), hundred)
    returns_rate = m.div(m.parse(i.returns_pct), hundred)
    marketing_total = m.parse(i.marketing_ngn)
    fixed_total = m.parse(i.fixed_cost_ngn)
    stress = m.div(m.parse(i.stress_pct), hundred)

    scenarios = []
    for name, price_factor, cost_factor in [('Downside', one - stress, one + stress), ('Base', one, one), ('Upside', one + stress, one - stress)]:
        supplier = m.mul(m.mul(unit_cost, fx), cost_factor)
        freight = m.mul(m.div(freight_total, quantity), cost_factor)
        duty = m.mul(supplier + freight, duty_rate)
        tax = m.mul(supplier + freight + duty, tax_rate)
        landed = supplier + freight + duty + tax
        price = m.mul(selling_price, price_factor)
        fees = m.mul(price, fee_rate)
        returns = m.mul(price, returns_rate)
        marketing = m.div(marketing_total, quantity)
        overhead = m.div(fixed_total, quantity)
        contribution = price - landed - fees - returns - marketing - overhead
        before_fixed = price - landed - fees - returns
        scenarios.append({
            'name': name, 'supplier': m.money(supplier), 'freight': m.money(freight), 'duty': m.money(duty),
            'import_tax': m.money(tax), 'landed_cost': m.money(landed), 'price': m.money(price),
            'fees': m.money(fees), 'returns': m.money(returns), 'marketing': m.money(marketing), 'overhead': m.money(overhead),
            'contribution': m.money(contribution), 'margin_pct': m.money(m.mul(m.div(contribution, price), hundred)),
            'cash_required': m.money(m.mul(landed, quantity) + marketing_total + fixed_total),
            'break_even_cac': m.money(max(0, before_fixed - overhead)),
            'break_even_units': m.ceil_units(m.div(marketing_total + fixed_total, before_fixed)) if before_fixed > 0 else None,
        })
    return scenarios


SCENARIO_FORMULAS = {
    'unit-economics/1.0.0': _scenarios_v1_0_0,
    'unit-economics/1.1.0': _scenarios_v1_1_0,
}


#: What a reviewer's rejection says, in the assessment as well as on the screen.
REJECTED_BLOCKER = 'A reviewer rejected this product for import. Do not proceed.'
PROHIBITED_BLOCKER = 'The product is marked prohibited. Do not proceed.'


def _blockers(scenarios, inputs, evidence):
    """The advisory blockers. Identical in both threshold versions."""
    base, downside = scenarios[1], scenarios[0]
    blockers = []
    if not evidence['coverage']: blockers.append('Collect independent demand signals and destination-market evidence.')
    if not evidence['compliance_resolved']: blockers.append('Obtain a reviewed product classification and current import requirements.')
    if evidence['confidence'] < 70: blockers.append('Raise evidence confidence to at least 70 before committing inventory.')
    if base['margin_pct'] < 25: blockers.append('Raise base contribution margin to at least 25%.')
    if downside['margin_pct'] < 10: blockers.append('Keep downside contribution margin at or above 10%.')
    return blockers


def _decide(scenarios, evidence, blockers):
    """The evidence/economics ladder, once nothing has forced NO-GO."""
    base = scenarios[1]
    if evidence['confidence'] < 40 or not evidence['coverage']:
        return 'INSUFFICIENT EVIDENCE'
    if base['margin_pct'] < 15 or (evidence['overall'] is not None and evidence['overall'] < 55):
        return 'NO-GO'
    if not blockers and (evidence['overall'] or 0) >= 75:
        return 'GO'
    return 'WATCH'


def _gates_v1_0_0(scenarios, inputs, evidence):
    """The gates the pilot shipped: only the user's own dropdown can force NO-GO.

    Frozen. Assessments saved under this version replay to the values they were saved
    with, including the ones review finding R04 identifies as wrong.
    """
    blockers = _blockers(scenarios, inputs, evidence)
    if inputs.compliance == 'prohibited':
        return 'NO-GO', [PROHIBITED_BLOCKER] + blockers
    return _decide(scenarios, evidence, blockers), blockers


def _gates_v1_1_0(scenarios, inputs, evidence):
    """The current gates: a reviewer's rejection is authoritative (review finding R04).

    Under 1.0.0 the gate text said "a reviewer rejected this product for import; do not
    proceed" while the decision beside it read WATCH, because the rejection reached the
    screen but never reached the calculation — only the user re-stating it in a dropdown
    did. A reviewer decision is the authority on import readiness, so it decides here.

    `compliance_status` is the reviewed state from the review workflow, computed on the
    server. It is absent from assessments saved before this version, and a missing value
    means "no reviewer has rejected this", which reproduces 1.0.0 exactly.
    """
    blockers = _blockers(scenarios, inputs, evidence)
    if evidence.get('compliance_status') == 'rejected':
        return 'NO-GO', [REJECTED_BLOCKER] + blockers
    if inputs.compliance == 'prohibited':
        return 'NO-GO', [PROHIBITED_BLOCKER] + blockers
    return _decide(scenarios, evidence, blockers), blockers


GATE_RULES = {
    'decision-gates/1.0.0': _gates_v1_0_0,
    'decision-gates/1.1.0': _gates_v1_1_0,
}


def gates(scenarios, inputs, evidence, threshold_version: str = THRESHOLD_VERSION):
    """The decision gates for `threshold_version`."""
    if threshold_version not in GATE_RULES:
        raise ValueError(f'unknown threshold version: {threshold_version}')
    return GATE_RULES[threshold_version](scenarios, inputs, evidence)


def economics_summary(scenarios):
    """Whether the scenario pays for itself, judged without reference to evidence.

    Kept out of `calculate()` on purpose: that function's output is pinned by
    `formula_version`, and a replayed historical assessment must reproduce byte for byte.
    This is a separate reading of the same scenarios so a screen can say "the economics
    fail under these assumptions" *and* "demand is unverified" instead of collapsing both
    into one verdict (action plan T07).
    """
    base, downside = scenarios[1], scenarios[0]
    failures = []
    if base['contribution'] <= 0:
        failures.append('The unit does not cover its own costs under these assumptions.')
    if base['margin_pct'] < 15:
        failures.append('Base contribution margin is below the 15% floor.')
    elif base['margin_pct'] < 25:
        failures.append('Base contribution margin is below the 25% target.')
    if downside['margin_pct'] < 10:
        failures.append('Downside contribution margin is below 10%.')
    return {'base_margin_pct': base['margin_pct'], 'downside_margin_pct': downside['margin_pct'],
            'contribution': base['contribution'], 'break_even_units': base['break_even_units'],
            'viable': not failures, 'failures': failures, 'threshold_version': THRESHOLD_VERSION}


def calculate(inputs: Inputs, evidence=None, formula_version: str = FORMULA_VERSION,
              threshold_version: str = THRESHOLD_VERSION):
    """Scenarios and gates for `inputs`.

    `formula_version` and `threshold_version` exist so a stored assessment replays under
    the versions it was saved with — the arithmetic and the decision rules move
    independently. Callers producing a *new* assessment must leave both at the default.
    """
    if formula_version not in SCENARIO_FORMULAS:
        raise ValueError(f'unknown formula version: {formula_version}')
    evidence = {**NO_EVIDENCE, **(evidence or {})}
    scenarios = SCENARIO_FORMULAS[formula_version](inputs)
    decision, blockers = gates(scenarios, inputs, evidence, threshold_version)
    return {'formula_version': formula_version, 'truth_state': 'Calculated', 'currency': 'NGN', 'market': 'NG',
            'decision': decision, 'confidence': evidence['confidence'], 'observation_ids': evidence['observation_ids'],
            'blockers': blockers, 'scenarios': scenarios, 'inputs': inputs.model_dump(), 'input_truth_state': 'User input'}
